cmd/skunk/adapter/in/networkAdapter/networkAdapter.go

```diff
@@ -0,0 +1,250 @@
+	"github.com/scherzma/Skunk/cmd/skunk/adapter/in/tor"
+	"github.com/scherzma/Skunk/cmd/skunk/application/port/network"
+	"github.com/scherzma/Skunk/cmd/skunk/util"
+)
+
+// Constants for configuration of Tor
+const (
+	SocksPort            = "9055"
+	LocalPort            = "1111"
+	RemotePort           = "2222"
+	ReusePrivateKey      = true // reuse private key for constant onion address
+	DeleteDataDirOnClose = false
+	UseEmbedded          = true // use embedded tor process
+)
+
+var (
+	networkAdapter *NetworkAdapter
+	once           sync.Once
+)
+
+// NetworkAdapter connects the main logic to the tor peer network
+type NetworkAdapter struct {
+	subscriber network.NetworkObserver // subscriber observing network messages
+	peer       *peer.Peer
+	tor        *tor.Tor
+}
+
+func NewAdapter() *NetworkAdapter {
+	// singleton
+	once.Do(func() {
+		networkAdapter = &NetworkAdapter{}
+	})
+
+	return networkAdapter
+}
+
+func (n *NetworkAdapter) SubscribeToNetwork(observer network.NetworkObserver) error {
+	if n.subscriber == observer {
+		return fmt.Errorf("network adapter is already connected to observer: %v", observer)
+	}
+
+	if n.peer != nil || n.tor != nil {
+		return fmt.Errorf("network services are already running")
+	}
+
+	torInstance, onionService, err := startTor()
+	if err != nil {
+		return err
+	}
+
+	peerInstance, err := startPeer(onionService)
+	if err != nil {
+		return err
+	}
+
+	// begin asynchronously reading network messages.
+	go n.readNetworkMessages()
+
+	n.subscriber = observer
+	n.peer = peerInstance
+	n.tor = torInstance
+
+	// notify the subscriber that the network is now online and send the onion address
+	message := network.Message{
+		Id:              util.UUID(),
+		Timestamp:       util.CurrentTimeMillis(),
+		Content:         n.peer.Address,
+		FromUser:        "",
+		SenderAddress:   "",
+		ReceiverAddress: "",
+		ChatID:          "",
+		Operation:       network.NETWORK_ONLINE,
+	}
+	n.SendNetworkMessageToSubscriber(message)
+	return nil
+}
+
+// UnsubscribeFromNetwork stops all network services and unsubscribes the observer
+func (n *NetworkAdapter) UnsubscribeFromNetwork() error {
+	if n.subscriber == nil {
+		return fmt.Errorf("can't unsubscribe from nil")
+	}
+
+	if n.peer == nil {
+		return fmt.Errorf("peer network is nil")
+	}
+
+	if n.tor == nil {
+		return fmt.Errorf("tor network is nil")
+	}
+
+	// stop tor and peer services
+	stopTor(n.tor)
+	stopPeer(n.peer)
+
+	n.subscriber = nil
+	n.peer = nil
+	n.tor = nil
+	return nil
+}
+
+// SendMessageToNetworkPeer sends a message to a specified network peer
+func (n *NetworkAdapter) SendMessageToNetworkPeer(address string, message network.Message) error {
+	// connect to peer if not already connected
+	if !n.peer.IsConnectedTo(address) {
+		err := n.peer.Connect(address)
+		// if there is any error, we treat it as if the peer is offline
+		if err != nil {
+			// message subscriber that the peer is offline
+			message := network.Message{
```

Comment:
The network.Message struct will get changes in one of the next pull requests. This works for now, but will have to be adjusted in the future.